x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -290,117 +264,72 @@ static String getPolicyId(SnapshotInfo snapshotInfo) {
+                new GroupedActionListener<>(ActionListener.runAfter(ActionListener.map(listener, v -> null),
+                        () -> {
+                            TimeValue totalElapsedTime = TimeValue.timeValueNanos(nowNanoSupplier.getAsLong() - startTime);
+                            logger.debug("total elapsed time for deletion of [{}] snapshots: {}", deleted, totalElapsedTime);
+                            slmStats.deletionTime(totalElapsedTime);
+                        }), snapshotsToDelete.size());
         for (Map.Entry<String, List<SnapshotInfo>> entry : snapshotsToDelete.entrySet()) {
             String repo = entry.getKey();
             List<SnapshotInfo> snapshots = entry.getValue();
-            for (SnapshotInfo info : snapshots) {
-                final String policyId = getPolicyId(info);
-                final long deleteStartTime = nowNanoSupplier.getAsLong();
-                // TODO: Use snapshot multi-delete instead of this loop if all nodes in the cluster support it
-                //       i.e are newer or equal to SnapshotsService#MULTI_DELETE_VERSION
-                deleteSnapshot(policyId, repo, info.snapshotId(), slmStats, ActionListener.wrap(acknowledgedResponse -> {
-                    deleted.incrementAndGet();
-                    assert acknowledgedResponse.isAcknowledged();
-                    historyStore.putAsync(SnapshotHistoryItem.deletionSuccessRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo));
-                }, e -> {
-                    failed.incrementAndGet();
-                    try {
-                        final SnapshotHistoryItem result = SnapshotHistoryItem.deletionFailureRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo, e);
-                        historyStore.putAsync(result);
-                    } catch (IOException ex) {
-                        // This shouldn't happen unless there's an issue with serializing the original exception
-                        logger.error(new ParameterizedMessage(
-                            "failed to record snapshot deletion failure for snapshot lifecycle policy [{}]",
-                            policyId), ex);
-                    }
-                }));
-                // Check whether we have exceeded the maximum time allowed to spend deleting
-                // snapshots, if we have, short-circuit the rest of the deletions
-                long finishTime = nowNanoSupplier.getAsLong();
-                TimeValue deletionTime = TimeValue.timeValueNanos(finishTime - deleteStartTime);
-                logger.debug("elapsed time for deletion of [{}] snapshot: {}", info.snapshotId(), deletionTime);
-                TimeValue totalDeletionTime = TimeValue.timeValueNanos(finishTime - startTime);
-                if (totalDeletionTime.compareTo(maximumTime) > 0) {
```

Comment:
This is irrelevant now, the loop over the deletes doesn't block so we will never break the time limit practically.